New leaderboard score card design implementation

mk56-spn opened the pull request:
- Part of #22204 
- Should resolve https://github.com/ppy/osu/issues/9411
- [x] Depends on https://github.com/ppy/osu/pull/25093

Figma reference: https://www.figma.com/file/EjS27IBaZhc2qX97lFBUxN/Song-Select-5?type=design&node-id=1-1010&mode=design&t=WQyZkwmDC4JWz6An-0

More song select stuff.

https://github.com/ppy/osu/assets/22781491/697f3290-ef30-4b2c-820f-51d7909b8f4e

<details>
<summary>old</summary>

Couple of things to note when reviewing this: 

Something to note while reviewing this, due to the amount of restructuring that will be required to fit these into song select they have been PRed in isolation, a follow up PR, most likely by me if im allowed will go through the necessary refactor to fit these into a new leaderboard, with the appropriate height, shear, etc.

- I do not recommend comparing the implementation to the old one, since ive split the commits to be reviewed as if this was an entirely new component, since i found it to be probably the easiest way to understand it.

- As for the overlapping mods when the count exceeds the amount that can be housed inside the right side of the component they will collapse and overlap as per @peppy 's suggestion. 

- And final major thing to note, i opted to specify the widths of the statistics manually, as to avoid them differing in length across different scores, which looked horrible when placed in a leaderboard together.

![sssssss](https://user-images.githubusercontent.com/74463310/212744729-8171632e-db72-41d3-88e4-59595c901abd.png)

Header should be coming some time in the next few days as well as the new tooltip for this component.

</details>

mk56-spn committed:
Set up test scene and basic LeaderBoardScoreV2.cs shape

mk56-spn committed:
Add grid container with sections, add background colours and hover logic

mk56-spn committed:
setup up rank and score logic flow

mk56-spn committed:
Add right side content

mk56-spn committed:
Add flag and user name

mk56-spn committed:
add score component label and animate component

mk56-spn committed:
add statistics to main content

mk56-spn committed:
Adjust widths in statistic labels

mk56-spn committed:
Add shading to mod pills

mk56-spn committed:
Add back tooltip

mk56-spn committed:
Add tooltip to new mod pills

Joehuu commented on osu.Game/Online/Leaderboards/LeaderBoardScoreV2.cs:
```diff
@@ -0,0 +1,550 @@
+                    }
+                };
+
+                if (statisticInfo.Name == EditorSetupStrings.ComboColourPrefix.ToUpper())
```
Shouldn't use this string as it's a different context. Either way it should be "max combo" to conform with other places (i.e. just ignore what's on the design here).

mk56-spn replied:
gotcha

Joehuu commented on osu.Game/Online/Leaderboards/LeaderBoardScoreV2.cs:
```diff
@@ -0,0 +1,550 @@
+                        break;
+
+                    case HitResult.Great:
+                        appendToString("300/");
```
Also ignore the design here, judgements in lazer are not numbers.

mk56-spn replied:
alr

mk56-spn committed:
fix capitalisation issue in class name

Co-authored-by: Joseph Madamba <madamba.joehu@outlook.com>

mk56-spn committed:
Actually fix issue with naming of LeaderboardScoreV2.cs class

mk56-spn committed:
clean up linQ result formatting. Replace numbers with hitresult displaynames.

Make adjustments to statistics to allow them to work with autosizing

peppy commented:
So what happens for a mania score? Can you add one to the test scene please? I don't think information is going to fit due to more judgements.

When designing tests, please make sure to not just consider the scenario which works well, but intentionally look for the edge cases that need to be supported. Without this, the tests are not that useful.

mk56-spn committed:
Add mania score to leaderboard test scene

mk56-spn commented:
![zzz](https://user-images.githubusercontent.com/74463310/214018630-351aa4d9-8cbe-43f3-b925-9c002e464072.png)

added to the test, fwiw i had already guarded against this by making the container for the judgements autosize to the size of the judgement names line . although i feel perhaps it would look better if the judgements were centered underneath?

peppy commented:
@arflyte are you sure this is the best way to display this? it feels a bit.. half-assed? especially when there's more judgements, it's very hard to visually parse the mapping of judgement names to counts.

mk56-spn commented:
> @arflyte are you sure this is the best way to display this? it feels a bit.. half-assed? especially when there's more judgements, it's very hard to visually parse the mapping of judgement names to counts.

I'm for moving this to the tooltip. It's not information you really need at a first glance ( atleaet in my experience) and that way the tooltip isn't just an expanded mod panel? As well as displaying them in a much more comprehensible way

peppy commented:
Keep in mind that we already have this in the tooltip on `master`, but it looks bad. I think we've asked @arflyte for a design for that in the past but haven't gotten anything yet. I think it's supposed to popup a larger view of the score.

mk56-spn commented:
> Keep in mind that we already have this in the tooltip on `master`, but it looks bad. I think we've asked @arflyte for a design for that in the past but haven't gotten anything yet. I think it's supposed to popup a larger view of the score.

just to clarify im kind of unsure how i should take this, am i meant to pull something out of somewhere and see if it works or await a design from @arflyte

arflyte commented:
I've been looking into this for the past few weeks; it's part of the song select revision and mod display fix. I still have no final solution to this yet, but it's one of the higher priorities I'm working on right now.

mk56-spn commented:
> I've been looking into this for the past few weeks; it's part of the song select revision and mod display fix. I still have no final solution to this yet, but it's one of the higher priorities I'm working on right now.

yeah no hurries on my part, and thanks for keeping me informed

Joehuu committed:
Merge remote-tracking branch 'upstream/master' into pr/22237

Joehuu committed:
Fix/update score exporting method

Joehuu committed:
Nuke hit results display

Joehuu committed:
Add relative width slider to test

Joehuu committed:
Move drawable init properties to constructor

Joehuu committed:
Add user covers to centre content

Joehuu committed:
Add shadow to centre content

Done this way instead of edge effect because of 1px/dimming issues.

Joehuu committed:
Remove weird yellow background override on mods

To who's reading, don't follow figma designs all the time. In most cases, follow what other places are doing.

Joehuu committed:
Populate dates and add show animation on test

Joehuu committed:
Fix leaderboard score caching colour provider

Joehuu committed:
Use `AutoSizeAxes` instead of hardcoded `Size`

Joehuu committed:
Move test scores to a method and add `TestResources.CreateTestScoreInfo()`

Joehuu committed:
Add draw width statistic to test

Joehuu committed:
Use grid container for centre content

Joehuu committed:
Truncate name label and clean up positioning code

Also adds a test score with a long username.

Joehuu committed:
Rename `createRightSideContent()` to `createRightContent()`

Joehuu committed:
Fix rank label tooltip area

Joehuu committed:
Reduce indents of private container methods

Joehuu committed:
Fix user cover having shear

Joehuu committed:
Update right content to latest design

- Add more scenarios to test
- Future-proof mods display to not overflow

Joehuu committed:
Add experimental collapse content logic based on width

Joehuu committed:
Add set up steps to reinit drawables with a different relative width

Joehuu committed:
Fix statistics container showing for a brief moment on lower widths

Joehuu committed:
Move invalidation issue todo to tests

Joehuu commented:
Have updated the design and experimented with the layout to prevent overflow:

https://github.com/ppy/osu/assets/35318437/d930e00d-857e-4017-ba40-1ae11e100d7e

The max combo and accuracy are not in the tooltip yet, but there is no finalized design for that yet.

Here's a branch with the changes required to see in-game: https://github.com/ppy/osu/compare/master...Joehuu:osu:leaderboard-score-redesign-ingame

bdach commented:
- What's with the gradient on the text? I don't see that on design.
- What happens if you use non-standard rate on DT / HT? I imagine the answer is "the rate doesn't show", which is a regression from the current UX. The tiny switches need to be able to show DT / HT rate / other extended info if they're to be used in this way.
- The indicator of when the score was set is not vertically centred with the flag.

Joehuu committed:
Remove alternative total score display (colour gradient) for now

For simplicity and a future consideration for when the skinning portion is implemented.

Joehuu committed:
Fix date not aligning with flag

Joehuu committed:
Remove unused field

Joehuu committed:
Merge remote-tracking branch 'upstream/master' into pr/22237

Joehuu committed:
Modify some half time mods on test

For use after support of extended info on `ModSwitchTiny`.

Joehuu committed:
Revert changes to `ModSwitchTiny`

Joehuu commented:
- That was an alternative design, see right on figma for more content.
- Have implemented, but needs design consideration when the mod icons are contracted (i.e. should it not display?)
- Fixed by using `UseFullGlyphHeight`

peppy commented:
I kinda like the gradients, think we should probably push them out and see what the feedback is like?

peppy commented:
![CleanShot 2023-10-12 at 04 50 40](https://github.com/ppy/osu/assets/191335/1da28d49-f8ea-447b-9e07-dc365641a3ad)

I dunno what's happening here but I don't like it. It's like the box is getting thinner as it goes further to the right?

I'm also not too keen on this dark border thing here. I don't think it fits well with the rest of the card:

![CleanShot 2023-10-12 at 04 51 22](https://github.com/ppy/osu/assets/191335/d1a68b0a-a388-4897-aae1-919329cc5c52)

@arflyte please check on these.

arflyte commented:
Added more alternative designs

https://www.figma.com/file/9bpXaCQNLuYQlbtyzcs115/Song-Select-6?type=design&node-id=0%3A1&mode=design&t=lEv5wCZC7VRK0dYH-1

peppy committed:
Merge branch 'master' into LeaderBoardScore_clean

Joehuu committed:
Fix compile errors

Joehuu committed:
Remove dark border and update cover gradient

arflyte commented:
Hmm, I think having the entire background filled with triangles would work fine too. Haven't tried it but I think it would work.

peppy commented:
@Joehuu to clarify my issue, it looks like the left  element is trapezoiding horizontally:

![2024-05-02 09 57 56@2x](https://github.com/ppy/osu/assets/191335/23c08378-01da-45e5-a8d9-311faec9e07e)

I'd hope this wasn't a thing and it took up the full height.

Joehuu commented:
That was an oversight on my part in implementing that, the gradient in figma doesn't disappear that quickly.

> I'd hope this wasn't a thing and it took up the full height.

That would be easier, code/layout-wise. Might do that to keep things moving.

bdach commented:
> That was an oversight on my part in implementing that, the gradient in figma doesn't disappear that quickly.

I'm not sure that matters? Whatever gradient is put on there, it will have a similar effect, no?

Joehuu committed:
Remove total score border gradient

Joehuu commented:
When you see the figma design at a normal zoom, it does exhibit the same effect yes. Been looking at things too zoomed in.

Here's what it looks like now:

![Screenshot 2024-05-02 at 10 56 47 PM](https://github.com/ppy/osu/assets/35318437/04952df8-aa0a-4000-8805-6d5f23e12f80)

peppy commented:
@Joehuu i'm not against the fluid layout, but cases like this can't exist IMO:

![2024-05-03 14 16 58@2x](https://github.com/ppy/osu/assets/191335/f97ff255-2680-451d-899d-3bb5bf29ccc1)

![2024-05-03 14 17 27@2x](https://github.com/ppy/osu/assets/191335/5c78fbd3-df0c-4b8f-9dac-549688e86e12)

Basically, the display mode should be based on constant (widths?) rather than if certain information can fit or not. I'm not sure how practical this is to fix (I haven't checked the code yet), open to discussion.

peppy reviewed (commented):
As mentioned.

Joehuu committed:
Fix fluidity desync by not autosizing right (total score) content

Joehuu committed:
Fix broken avatar masking when hiding

Joehuu committed:
Lower username font size a bit

Joehuu commented:
Have fixed some of the issues mentioned here and discord.

Here's my v2 of the fluidity layout (haven't committed as it may be turned down):

<details>
<summary>Diff</summary>

```diff
diff --git a/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs b/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
index dea134b4d6..e68156a7ed 100644
--- a/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
+++ b/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
@@ -52,7 +52,11 @@ public partial class LeaderboardScoreV2 : OsuClickableContainer, IHasContextMenu
 
         private const float right_content_width = 180;
         private const float grade_width = 40;
-        private const float username_min_width = 100;
+        private const float username_min_width = 125;
+        private const float statistics_regular_min_width = 175;
+        private const float statistics_compact_min_width = 100;
+        private const float rank_label_width = 65;
+        private const float rank_label_visibility_width_cutoff = rank_label_width + height + username_min_width + statistics_regular_min_width + right_content_width;
 
         private readonly ScoreInfo score;
 
@@ -101,9 +105,9 @@ public partial class LeaderboardScoreV2 : OsuClickableContainer, IHasContextMenu
         private Drawable scoreRank = null!;
         private Box totalScoreBackground = null!;
 
-        private Container centreContent = null!;
-        private FillFlowContainer usernameAndFlagContainer = null!;
         private FillFlowContainer statisticsContainer = null!;
+        private RankLabel rankLabel = null!;
+        private Container rankLabelOverlay = null!;
 
         public ITooltip<ScoreInfo> GetCustomTooltip() => new LeaderboardScoreTooltip();
         public virtual ScoreInfo TooltipContent => score;
@@ -151,7 +155,7 @@ private void load()
                         RelativeSizeAxes = Axes.Both,
                         ColumnDimensions = new[]
                         {
-                            new Dimension(GridSizeMode.Absolute, 65),
+                            new Dimension(GridSizeMode.AutoSize),
                             new Dimension(),
                             new Dimension(GridSizeMode.Absolute, right_content_width),
                         },
@@ -159,8 +163,17 @@ private void load()
                         {
                             new Drawable[]
                             {
-                                new RankLabel(rank) { Shear = -shear },
-                                centreContent = createCentreContent(user),
+                                new Container
+                                {
+                                    AutoSizeAxes = Axes.X,
+                                    RelativeSizeAxes = Axes.Y,
+                                    Child = rankLabel = new RankLabel(rank)
+                                    {
+                                        Width = rank_label_width,
+                                        RelativeSizeAxes = Axes.Y,
+                                    },
+                                },
+                                createCentreContent(user),
                                 createRightContent()
                             }
                         }
@@ -214,21 +227,43 @@ private void load()
                                 AutoSizeAxes = Axes.Both,
                                 CornerRadius = corner_radius,
                                 Masking = true,
-                                Child = avatar = new DelayedLoadWrapper(
-                                    innerAvatar = new ClickableAvatar(user)
+                                Children = new[]
+                                {
+                                    avatar = new DelayedLoadWrapper(
+                                        innerAvatar = new ClickableAvatar(user)
+                                        {
+                                            Anchor = Anchor.Centre,
+                                            Origin = Anchor.Centre,
+                                            Scale = new Vector2(1.1f),
+                                            Shear = -shear,
+                                            RelativeSizeAxes = Axes.Both,
+                                        })
+                                    {
+                                        RelativeSizeAxes = Axes.None,
+                                        Size = new Vector2(height)
+                                    },
+                                    rankLabelOverlay = new Container
                                     {
-                                        Anchor = Anchor.Centre,
-                                        Origin = Anchor.Centre,
-                                        Scale = new Vector2(1.1f),
-                                        Shear = -shear,
                                         RelativeSizeAxes = Axes.Both,
-                                    })
-                                {
-                                    RelativeSizeAxes = Axes.None,
-                                    Size = new Vector2(height)
+                                        Alpha = 0,
+                                        Children = new Drawable[]
+                                        {
+                                            new Box
+                                            {
+                                                RelativeSizeAxes = Axes.Both,
+                                                Colour = Colour4.Black.Opacity(0.5f),
+                                            },
+                                            new RankLabel(rank)
+                                            {
+                                                AutoSizeAxes = Axes.Both,
+                                                Anchor = Anchor.Centre,
+                                                Origin = Anchor.Centre,
+                                            },
+                                        }
+                                    }
                                 },
                             },
-                            usernameAndFlagContainer = new FillFlowContainer
+                            new FillFlowContainer
                             {
                                 Anchor = Anchor.CentreLeft,
                                 Origin = Anchor.CentreLeft,
@@ -279,7 +314,7 @@ private void load()
                                 {
                                     Name = @"Statistics container",
                                     Padding = new MarginPadding { Right = 40 },
-                                    Spacing = new Vector2(25),
+                                    Spacing = new Vector2(25, 0),
                                     Shear = -shear,
                                     Anchor = Anchor.CentreRight,
                                     Origin = Anchor.CentreRight,
@@ -287,6 +322,8 @@ private void load()
                                     Direction = FillDirection.Horizontal,
                                     Children = statisticsLabels,
                                     Alpha = 0,
+                                    LayoutEasing = Easing.OutQuint,
+                                    LayoutDuration = transition_duration,
                                 }
                             }
                         }
@@ -483,6 +520,11 @@ private void updateState()
             foreground.FadeColour(IsHovered ? foregroundColour.Lighten(0.2f) : foregroundColour, transition_duration, Easing.OutQuint);
             background.FadeColour(IsHovered ? backgroundColour.Lighten(0.2f) : backgroundColour, transition_duration, Easing.OutQuint);
             totalScoreBackground.FadeColour(IsHovered ? lightenedGradient : totalScoreBackgroundGradient, transition_duration, Easing.OutQuint);
+
+            if (DrawWidth < rank_label_visibility_width_cutoff && IsHovered)
+                rankLabelOverlay.FadeIn(transition_duration, Easing.OutQuint);
+            else
+                rankLabelOverlay.FadeOut(transition_duration, Easing.OutQuint);
         }
 
         protected override bool OnInvalidate(Invalidation invalidation, InvalidationSource source)
@@ -490,22 +532,27 @@ protected override bool OnInvalidate(Invalidation invalidation, InvalidationSour
             Scheduler.AddOnce(() =>
             {
                 // when width decreases
-                // - hide statistics, then
-                // - hide avatar, then
-                // - hide user and flag and show avatar again
-
-                if (centreContent.DrawWidth >= height + username_min_width || centreContent.DrawWidth < username_min_width)
-                    avatar.FadeIn(transition_duration, Easing.OutQuint).MoveToX(0, transition_duration, Easing.OutQuint);
-                else
-                    avatar.FadeOut(transition_duration, Easing.OutQuint).MoveToX(-avatar.DrawWidth, transition_duration, Easing.OutQuint);
+                // - hide rank and show rank overlay on avatar when hovered
+                // - compact statistics, then
+                // - hide statistics
 
-                if (centreContent.DrawWidth >= username_min_width)
-                    usernameAndFlagContainer.FadeIn(transition_duration, Easing.OutQuint).MoveToX(0, transition_duration, Easing.OutQuint);
+                if (DrawWidth >= rank_label_visibility_width_cutoff)
+                    rankLabel.FadeIn(transition_duration, Easing.OutQuint).MoveToX(0, transition_duration, Easing.OutQuint);
                 else
-                    usernameAndFlagContainer.FadeOut(transition_duration, Easing.OutQuint).MoveToX(usernameAndFlagContainer.DrawWidth, transition_duration, Easing.OutQuint);
+                    rankLabel.FadeOut(transition_duration, Easing.OutQuint).MoveToX(-rankLabel.DrawWidth, transition_duration, Easing.OutQuint);
 
-                if (centreContent.DrawWidth >= height + statisticsContainer.DrawWidth + username_min_width)
+                if (DrawWidth >= height + username_min_width + statistics_regular_min_width + right_content_width)
+                {
                     statisticsContainer.FadeIn(transition_duration, Easing.OutQuint).MoveToX(0, transition_duration, Easing.OutQuint);
+                    statisticsContainer.Direction = FillDirection.Horizontal;
+                    statisticsContainer.ScaleTo(1, transition_duration, Easing.OutQuint);
+                }
+                else if (DrawWidth >= height + username_min_width + statistics_compact_min_width + right_content_width)
+                {
+                    statisticsContainer.FadeIn(transition_duration, Easing.OutQuint).MoveToX(0, transition_duration, Easing.OutQuint);
+                    statisticsContainer.Direction = FillDirection.Vertical;
+                    statisticsContainer.ScaleTo(0.8f, transition_duration, Easing.OutQuint);
+                }
                 else
                     statisticsContainer.FadeOut(transition_duration, Easing.OutQuint).MoveToX(statisticsContainer.DrawWidth, transition_duration, Easing.OutQuint);
             });
@@ -577,15 +624,14 @@ private partial class RankLabel : Container, IHasTooltip
         {
             public RankLabel(int? rank)
             {
-                AutoSizeAxes = Axes.Both;
-                Anchor = Anchor.Centre;
-                Origin = Anchor.Centre;
-
                 if (rank >= 1000)
                     TooltipText = $"#{rank:N0}";
 
                 Child = new OsuSpriteText
                 {
+                    Shear = -shear,
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
                     Font = OsuFont.GetFont(size: 20, weight: FontWeight.SemiBold, italics: true),
                     Text = rank == null ? "-" : rank.Value.FormatRank().Insert(0, "#")
                 };

```

</details>

https://github.com/ppy/osu/assets/35318437/e90666d6-bee4-4996-b3ac-2f414e1673ea

- Increased minimum username width
- Considered rank to be not as important, so it gets hidden first (and overlaid on the avatar when hovering score)
	- leaderboard always show 1-50 for now, can be reconsidered if needed/wanted for future relative rankings
	- user rank can replace your username in personal best like stable
	- avatar should be always shown now (for rank and profile viewing)
- Statistics have a compacted state before it gets hidden

Haven't done dynamic font size changing as suggested as I feel it's counterintuitive (at least when increasing ui scale). Padding can be done last to squeeze every pixel before something gets hidden. Just want to get the layout / fluidity states right first.

peppy commented:
Seems better. Maybe too many different view stages/modes but hard to say without seeing this in its final usage.

Joehuu committed:
Improve fluidity states

Joehuu committed:
Apply missed changes from old leaderboard score

See:
- https://github.com/ppy/osu/commit/d11e56b8bb6ade5c4f6e47de5fa288a909f6bc66
- https://github.com/ppy/osu/commit/7d74d84e6c24e6938d27fae0d7322e113df4be94
- https://github.com/ppy/osu/commit/07f9f5c6d842d3c7c564f96576682b6fb54c50b4

Givikap120 commented:
pp seems like a more important metric than combo, not even talking about judgements
people are usually not interested in exact amount of 100s and 50s, they're only looking on accuracy and misscount

smoogipoo commented:
pp is not relevant for a score leaderboard. Save that for a different mode or something.

peppy commented:
I dunno about that. Looking at this now it *is* weird that pp isn't shown, especially since we went out of our way to start showing it on web beatmap  leaderboards to stop majority of users from installing third party browser extensions to add it.

Feels like we should be showing it at least in the most expanded mode. Or in the tooltip.

frenzibyte commented:
Is it only me who really dislikes having mods display both in contracted and expanded state depends on mods count?

![CleanShot 2024-05-14 at 07 21 26](https://github.com/ppy/osu/assets/22781491/07e65f64-2cd3-42dc-aa55-28d034046cda)

What would everyone's opinion be about these two proposals:
 1. Compute spacing based on number of mods:

    ![CleanShot 2024-05-14 at 08 53 24](https://github.com/ppy/osu/assets/22781491/3a48d454-1fd8-442e-8e42-8378c31fa5fc)

 2. Use constant contracted spacing and cut off at 7+ mods:

    ![CleanShot 2024-05-14 at 08 50 41](https://github.com/ppy/osu/assets/22781491/9bb94c7a-57f5-4362-a033-402ecb05d000)

<details>
<summary>diff for number 1</summary>

```diff
diff --git a/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs b/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
index c5f96d1568..aca61c4d9a 100644
--- a/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
+++ b/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
@@ -154,19 +154,20 @@ private static ScoreInfo[] getTestScores()
                 }
             };
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_EXPANDED; i++)
+            for (int i = 0; i < 3; i++)
                 scores[0].Mods = scores[0].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : halfTime }).ToArray();
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_EXPANDED + 1; i++)
+            for (int i = 0; i < 4; i++)
                 scores[1].Mods = scores[1].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : new OsuModHalfTime() }).ToArray();
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_CONTRACTED; i++)
+            for (int i = 0; i < 5; i++)
                 scores[2].Mods = scores[2].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : halfTime }).ToArray();
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_CONTRACTED + 1; i++)
+            for (int i = 0; i < 6; i++)
                 scores[3].Mods = scores[3].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : new OsuModHalfTime() }).ToArray();
 
-            scores[4].Mods = scores[4].BeatmapInfo!.Ruleset.CreateInstance().CreateAllMods().ToArray();
+            for (int i = 0; i < 5; i++)
+                scores[4].Mods = scores[4].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : new OsuModHalfTime() }).ToArray();
 
             return scores;
         }
diff --git a/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs b/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
index b9ae3bb20e..153f000836 100644
--- a/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
+++ b/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
@@ -182,8 +182,21 @@ private void load()
             };
 
             innerAvatar.OnLoadComplete += d => d.FadeInFromZero(200);
+        }
+
+        private const float mod_scale = 0.375f;
+
+        protected override void Update()
+        {
+            base.Update();
+
+            float modsContainerWidth = ModSwitchTiny.WIDTH * mod_scale * modsContainer.Children.Count;
+            float requiredSpacing = Math.Min((modsContainer.DrawWidth - modsContainerWidth) / (modsContainer.Children.Count - 1), 1f);
 
-            modsContainer.Spacing = new Vector2(modsContainer.Children.Count > MAX_MODS_EXPANDED ? -20 : 2, 0);
+            if (modsContainer.Children.Count == 0)
+                requiredSpacing = 0;
+
+            modsContainer.Spacing = new Vector2(requiredSpacing, 0);
             modsContainer.Padding = new MarginPadding { Top = modsContainer.Children.Count > 0 ? 4 : 0 };
         }
 
@@ -432,14 +445,20 @@ private void load()
                                         Current = scoreManager.GetBindableTotalScoreString(score),
                                         Font = OsuFont.GetFont(size: 30, weight: FontWeight.Light),
                                     },
-                                    modsContainer = new FillFlowContainer<ColouredModSwitchTiny>
+                                    modsContainer = new ReverseChildIDFillFlowContainer<ColouredModSwitchTiny>
                                     {
                                         Anchor = Anchor.TopRight,
                                         Origin = Anchor.TopRight,
                                         Shear = -shear,
-                                        AutoSizeAxes = Axes.Both,
+                                        AutoSizeAxes = Axes.Y,
+                                        Width = ModSwitchTiny.WIDTH * mod_scale * 4 + 4 * 3,
                                         Direction = FillDirection.Horizontal,
-                                        ChildrenEnumerable = score.Mods.AsOrdered().Select(mod => new ColouredModSwitchTiny(mod) { Scale = new Vector2(0.375f) })
+                                        ChildrenEnumerable = score.Mods.AsOrdered().Select(mod => new ColouredModSwitchTiny(mod)
+                                        {
+                                            Anchor = Anchor.TopRight,
+                                            Origin = Anchor.TopRight,
+                                            Scale = new Vector2(mod_scale)
+                                        })
                                     },
                                     modsCounter = new OsuSpriteText
                                     {
diff --git a/osu.Game/Rulesets/UI/ModSwitchTiny.cs b/osu.Game/Rulesets/UI/ModSwitchTiny.cs
index 4d50e702af..4a3bc9e31b 100644
--- a/osu.Game/Rulesets/UI/ModSwitchTiny.cs
+++ b/osu.Game/Rulesets/UI/ModSwitchTiny.cs
@@ -22,7 +22,7 @@ public partial class ModSwitchTiny : CompositeDrawable
         public BindableBool Active { get; } = new BindableBool();
 
         public const float DEFAULT_HEIGHT = 30;
-        private const float width = 73;
+        public const float WIDTH = 73;
 
         protected readonly IMod Mod;
         private readonly bool showExtendedInformation;
@@ -56,7 +56,7 @@ public ModSwitchTiny(IMod mod, bool showExtendedInformation = false)
                     Width = 100 + DEFAULT_HEIGHT / 2,
                     RelativeSizeAxes = Axes.Y,
                     Masking = true,
-                    X = width,
+                    X = WIDTH,
                     Margin = new MarginPadding { Left = -DEFAULT_HEIGHT },
                     Children = new Drawable[]
                     {
@@ -77,7 +77,7 @@ public ModSwitchTiny(IMod mod, bool showExtendedInformation = false)
                 },
                 new CircularContainer
                 {
-                    Width = width,
+                    Width = WIDTH,
                     RelativeSizeAxes = Axes.Y,
                     Masking = true,
                     Children = new Drawable[]
```

</details>

<details>
<summary>diff for number 2</summary>

```diff
diff --git a/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs b/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
index c5f96d1568..aca61c4d9a 100644
--- a/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
+++ b/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
@@ -154,19 +154,20 @@ private static ScoreInfo[] getTestScores()
                 }
             };
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_EXPANDED; i++)
+            for (int i = 0; i < 3; i++)
                 scores[0].Mods = scores[0].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : halfTime }).ToArray();
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_EXPANDED + 1; i++)
+            for (int i = 0; i < 4; i++)
                 scores[1].Mods = scores[1].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : new OsuModHalfTime() }).ToArray();
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_CONTRACTED; i++)
+            for (int i = 0; i < 5; i++)
                 scores[2].Mods = scores[2].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : halfTime }).ToArray();
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_CONTRACTED + 1; i++)
+            for (int i = 0; i < 6; i++)
                 scores[3].Mods = scores[3].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : new OsuModHalfTime() }).ToArray();
 
-            scores[4].Mods = scores[4].BeatmapInfo!.Ruleset.CreateInstance().CreateAllMods().ToArray();
+            for (int i = 0; i < 5; i++)
+                scores[4].Mods = scores[4].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : new OsuModHalfTime() }).ToArray();
 
             return scores;
         }
diff --git a/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs b/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
index b9ae3bb20e..d789d96151 100644
--- a/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
+++ b/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
@@ -41,14 +41,9 @@ namespace osu.Game.Online.Leaderboards
     public partial class LeaderboardScoreV2 : OsuClickableContainer, IHasContextMenu, IHasCustomTooltip<ScoreInfo>
     {
         /// <summary>
-        /// The maximum number of mods when contracted until the mods display width exceeds the <see cref="right_content_width"/>.
+        /// The maximum number of mods until the mods display width exceeds the <see cref="right_content_width"/>.
         /// </summary>
-        public const int MAX_MODS_CONTRACTED = 13;
-
-        /// <summary>
-        /// The maximum number of mods when expanded until the mods display width exceeds the <see cref="right_content_width"/>.
-        /// </summary>
-        public const int MAX_MODS_EXPANDED = 4;
+        public const int MAX_MODS = 6;
 
         private const float right_content_width = 180;
         private const float grade_width = 40;
@@ -183,7 +178,6 @@ private void load()
 
             innerAvatar.OnLoadComplete += d => d.FadeInFromZero(200);
 
-            modsContainer.Spacing = new Vector2(modsContainer.Children.Count > MAX_MODS_EXPANDED ? -20 : 2, 0);
             modsContainer.Padding = new MarginPadding { Top = modsContainer.Children.Count > 0 ? 4 : 0 };
         }
 
@@ -439,6 +433,7 @@ private void load()
                                         Shear = -shear,
                                         AutoSizeAxes = Axes.Both,
                                         Direction = FillDirection.Horizontal,
+                                        Spacing = new Vector2(-10, 0f),
                                         ChildrenEnumerable = score.Mods.AsOrdered().Select(mod => new ColouredModSwitchTiny(mod) { Scale = new Vector2(0.375f) })
                                     },
                                     modsCounter = new OsuSpriteText
@@ -492,7 +487,7 @@ public override void Show()
 
                     using (BeginDelayedSequence(50))
                     {
-                        Drawable modsDrawable = score.Mods.Length > MAX_MODS_CONTRACTED ? modsCounter : modsContainer;
+                        Drawable modsDrawable = score.Mods.Length > MAX_MODS ? modsCounter : modsContainer;
                         var drawables = new[] { flagBadgeAndDateContainer, modsDrawable }.Concat(statisticsLabels).ToArray();
                         for (int i = 0; i < drawables.Length; i++)
                             drawables[i].FadeIn(100 + i * 50);
```

</details>

bdach commented:
If you ask me, once the acronyms are not clearly readable, you might as well not bother showing the pills and start showing the count alone. So I'd probably just cap at 4 mods with no overlap and call it a day.

frenzibyte commented:
That is also a viable option as long as we agree we don't need to show more than 4 mods, I thought the point of going up to 7 mods is to resolve some recurring issues we faced with regards to overflow of mods, but I don't know if that's really the case or just an arbitrary change in this PR.

frenzibyte commented:
Upon taking a peek at some of the leaderboards, I think we should at least support showing up to 5 mods for the initial push of the screen, to handle the famous mod combination DT FL HR HD combined with the Classic mod. This is definitely going to show up in a lot of scores and just displaying a text saying "5 mods" feels like it misrepresents the score (as these combinations are quite common), let alone showing them in a contracted state.

Final proposal for the time being:

![CleanShot 2024-05-14 at 14 31 28](https://github.com/ppy/osu/assets/22781491/1d90b733-b2ca-4f3f-96ae-0a726f7fcca6)

Keeps design consistent and avoids mods overlapping each other with the "contracted" state (I think contraction is best suited for icon-based mods, which we have seemingly moved away from).

<details>
<summary>diff</summary>

```diff
diff --git a/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs b/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
index c5f96d1568..321f443a97 100644
--- a/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
+++ b/osu.Game.Tests/Visual/SongSelect/TestSceneLeaderboardScoreV2.cs
@@ -154,19 +154,11 @@ private static ScoreInfo[] getTestScores()
                 }
             };
 
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_EXPANDED; i++)
-                scores[0].Mods = scores[0].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : halfTime }).ToArray();
-
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_EXPANDED + 1; i++)
-                scores[1].Mods = scores[1].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : new OsuModHalfTime() }).ToArray();
-
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_CONTRACTED; i++)
-                scores[2].Mods = scores[2].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : halfTime }).ToArray();
-
-            for (int i = 0; i < LeaderboardScoreV2.MAX_MODS_CONTRACTED + 1; i++)
-                scores[3].Mods = scores[3].Mods.Concat(new Mod[] { i % 2 == 0 ? new OsuModHidden() : new OsuModHalfTime() }).ToArray();
-
-            scores[4].Mods = scores[4].BeatmapInfo!.Ruleset.CreateInstance().CreateAllMods().ToArray();
+            scores[0].Mods = new Mod[] { new OsuModDoubleTime(), new OsuModHidden() };
+            scores[1].Mods = new Mod[] { new OsuModDoubleTime(), new OsuModHidden(), new OsuModFlashlight() };
+            scores[2].Mods = new Mod[] { new OsuModDoubleTime(), new OsuModHidden(), new OsuModFlashlight(), new OsuModHardRock() };
+            scores[3].Mods = new Mod[] { new OsuModDoubleTime(), new OsuModHidden(), new OsuModFlashlight(), new OsuModHardRock(), new OsuModClassic() };
+            scores[4].Mods = new Mod[] { new OsuModDoubleTime(), new OsuModHidden(), new OsuModFlashlight(), new OsuModHardRock(), new OsuModClassic(), new OsuModDifficultyAdjust() };
 
             return scores;
         }
diff --git a/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs b/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
index b9ae3bb20e..862eae10e7 100644
--- a/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
+++ b/osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs
@@ -41,16 +41,11 @@ namespace osu.Game.Online.Leaderboards
     public partial class LeaderboardScoreV2 : OsuClickableContainer, IHasContextMenu, IHasCustomTooltip<ScoreInfo>
     {
         /// <summary>
-        /// The maximum number of mods when contracted until the mods display width exceeds the <see cref="right_content_width"/>.
+        /// The maximum number of mods until the mods display width exceeds the <see cref="right_content_width"/>.
         /// </summary>
-        public const int MAX_MODS_CONTRACTED = 13;
+        public const int MAX_MODS = 5;
 
-        /// <summary>
-        /// The maximum number of mods when expanded until the mods display width exceeds the <see cref="right_content_width"/>.
-        /// </summary>
-        public const int MAX_MODS_EXPANDED = 4;
-
-        private const float right_content_width = 180;
+        private const float right_content_width = 210;
         private const float grade_width = 40;
         private const float username_min_width = 125;
         private const float statistics_regular_min_width = 175;
@@ -182,8 +177,6 @@ private void load()
             };
 
             innerAvatar.OnLoadComplete += d => d.FadeInFromZero(200);
-
-            modsContainer.Spacing = new Vector2(modsContainer.Children.Count > MAX_MODS_EXPANDED ? -20 : 2, 0);
             modsContainer.Padding = new MarginPadding { Top = modsContainer.Children.Count > 0 ? 4 : 0 };
         }
 
@@ -439,6 +432,7 @@ private void load()
                                         Shear = -shear,
                                         AutoSizeAxes = Axes.Both,
                                         Direction = FillDirection.Horizontal,
+                                        Spacing = new Vector2(2f, 0f),
                                         ChildrenEnumerable = score.Mods.AsOrdered().Select(mod => new ColouredModSwitchTiny(mod) { Scale = new Vector2(0.375f) })
                                     },
                                     modsCounter = new OsuSpriteText
@@ -492,7 +486,7 @@ public override void Show()
 
                     using (BeginDelayedSequence(50))
                     {
-                        Drawable modsDrawable = score.Mods.Length > MAX_MODS_CONTRACTED ? modsCounter : modsContainer;
+                        Drawable modsDrawable = score.Mods.Length > MAX_MODS ? modsCounter : modsContainer;
                         var drawables = new[] { flagBadgeAndDateContainer, modsDrawable }.Concat(statisticsLabels).ToArray();
                         for (int i = 0; i < drawables.Length; i++)
                             drawables[i].FadeIn(100 + i * 50);
```

</summary>

frenzibyte commented on osu.Game/Online/Leaderboards/LeaderboardScoreV2.cs:
```diff
@@ -0,0 +1,686 @@
+                                        Direction = FillDirection.Horizontal,
+                                        ChildrenEnumerable = score.Mods.AsOrdered().Select(mod => new ColouredModSwitchTiny(mod) { Scale = new Vector2(0.375f) })
+                                    },
+                                    modsCounter = new OsuSpriteText
+                                    {
+                                        Anchor = Anchor.TopRight,
+                                        Origin = Anchor.TopRight,
+                                        Shear = -shear,
+                                        Text = $"{score.Mods.Length} mods",
+                                        Alpha = 0,
+                                    }
```
I have [stirred up](https://discord.com/channels/188630481301012481/188630652340404224/1239900324915974255) some discussion on whether this can be replaced by an indicator next to an arbitrary subset of mods, and it was concluded that it's better. For example (design is arbitrary):

![image](https://github.com/ppy/osu/assets/22781491/dba7b422-1421-4b21-84f9-02336e861d31)

Currently [waiting on response](https://discord.com/channels/188630481301012481/188630652340404224/1239908144482881546) from @arflyte for a better design of this indicator, unless it's fine to proceed with the current one (I definitely need to rewrite the code because I faked up a `Mod` to achieve this).

frenzibyte replied:
Added in https://github.com/ppy/osu/pull/22237/commits/59553780040dbf82c48b7bd675f0c64ba97194bd

Givikap120 commented:
I think that it's much more readable to use an icons instead of acronyms
Especially considering new players existing

frenzibyte committed:
Make scores slanted in test scene

frenzibyte committed:
Remove expanded/contracted states and limit to 5 mods

Also adjusts right content width to contain those 5 mods. Not sure how to handle the extra space in the score though...to be dealt with later.

frenzibyte committed:
Replace "X mods" text with a pill indicator

frenzibyte committed:
Adjust right content width based on scoring mode

frenzibyte committed:
Move to `SelectV2` namespace

frenzibyte commented:
I've changed the mod display as proposed in https://github.com/ppy/osu/pull/22237#issuecomment-2109985797 & https://github.com/ppy/osu/pull/22237#discussion_r1599917577. I've also made the width of the right content change based on scoring mode, to handle marathon beatmaps where scores can go up to 9 digits (e.g. https://osu.ppy.sh/beatmapsets/972#osu/9007).

Preview:

https://github.com/ppy/osu/assets/22781491/697f3290-ef30-4b2c-820f-51d7909b8f4e

This looks to be in a good state to me.

frenzibyte committed:
Move toggle step to `SetUpSteps`

Joehuu committed:
Follow slanted flow logic precedent in test

See `ModSelectOverlay` components.

Joehuu commented:
I've pushed a commit that follows precedent to the slanted flow logic (mod overlay components use `Shear = Vector2.Zero` while the parent uses `Shear = {SomeComponent}.SHEAR`).

When restarting the test, I thought the right content width logic was broken, but it was just changing the scoring mode initially. Not sure how to prevent that step from happening initially.

Recent changes look good to me otherwise.

peppy committed:
Merge branch 'master' into LeaderBoardScore_clean

Givikap120 commented:
> I think that it's much more readable to use an icons instead of acronyms Especially considering new players existing

can you please explain me why acronyms?

frenzibyte committed:
Merge branch 'master' into LeaderBoardScore_clean

frenzibyte committed:
Use `OsuGame.SHEAR`

developomp commented:
:tada:
